app/bundles/CoreBundle/Service/GlobalSearch.php

```diff
@@ -0,0 +1,135 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Mautic\CoreBundle\Service;
+
+use Mautic\ChannelBundle\Model\MessageModel;
+use Mautic\CoreBundle\Event\GlobalSearchEvent;
+use Mautic\CoreBundle\Helper\UserHelper;
+use Mautic\CoreBundle\Security\Permissions\CorePermissions;
+use Twig\Environment;
+
+class GlobalSearch
+{
+    public function __construct(
+        private CorePermissions $security,
+        private UserHelper $userHelper,
+        private Environment $twig,
+    ) {
+    }
+
+    /**
+     * @param object|MessageModel $model
+     *
+     * @return array<int|string, int|string>
+     */
+    public function performSearch(
+        string $searchString,
+        mixed $model,
+        string $template,
+        bool $skipPermissions = false,
+    ): array {
+        $permissions = $this->getPermissions($model->getPermissionBase(), $skipPermissions);
+
+        if (!$permissions['viewown'] && !$permissions['viewother']) {
+            return [];
+        }
+
+        $filter = ['string' => $searchString, 'force' => []];
+
+        if (!$permissions['viewother']) {
+            $filter['force'][] = [
+                'column' => 'createdBy',
+                'expr'   => 'eq',
+                'value'  => $this->userHelper->getUser()->getId(),
+            ];
+        }
+
+        $results = $model->getEntities([
+            'filter'           => $filter,
+            'start'            => 0,
+            'limit'            => GlobalSearchEvent::RESULTS_LIMIT,
+            'ignore_paginator' => true,
+            'with_total_count' => true,
+        ]);
+
+        return $this->processResults($results, $searchString, $template, $permissions);
+    }
+
+    /**
+     * Process search results and render templates.
+     *
+     * @param array<mixed>         $results
+     * @param array<string, bool> $permissions
+     *
+     * @return array<int|string, mixed>
+     */
+    private function processResults(
+        array $results,
+        string $searchString,
+        string $template,
+        array $permissions,
+    ): array {
+        $count = $results['count'] ? (int) $results['count'] : 0;
```

Comment:
It would be really nice if the `$results` would be a collection object that would hold the array of results and the count. Working with arrays this way is a pain.